### PR TITLE
feat: add hook to access current route info

### PR DESCRIPTION
**Pull request by EvanBacon**
- At any given time, it's hard to know what the current virtual URL is. 
- This PR adds: 
  - ~~`query`, `pathname`, and `asPath` to `useLink` (currently modeled after Next.js)~~.
  - A new hook `useHref` which returns `{ href: string, pathname: string, query: Record<string, any> }` based on the current state of the virtual navigator.
- This PR also moves `pathname` and `stateName` out of the custom Layout context in favor of `useHref`.
- Navigation state can now be observed via `RootContainer.useState()`.

### CODE REVIEW

**Comment by nandorojo** [comment]
What do you think of splitting this into a separate hook? 

This PR looks like it will re-render all consumers of `useLink` whenever the root navigation state's path changes. I think the Next.js design of pairing `query`/`asPath` into `useRouter()` was a mistake for this reason.

React Navigation separates `useRoute` from `useNavigation`. I think this makes more sense.

Also, just for reference, here's a similar discussion about `asPath` on the solito repo: https://github.com/nandorojo/solito/pull/45#issuecomment-1094305867

**Comment by EvanBacon** [comment]
@nandorojo good point, I split the route info into `useHref` which also re-renders much less now. `asPath` has been renamed to `href`, but not married to any of these names.

File: packages/expo-router/src/link/useHref.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
+import { LinkingContext } from "@react-navigation/native";
+import React from "react";
+
+import { RootContainer } from "../ContextNavigationContainer";
+import getPathFromState, { State } from "../fork/getPathFromState";
+
+// These values are based on Next.js router
```
Comment (EvanBacon) [inline packages/expo-router/src/link/useHref.ts:7]:
```suggestion
```

**Comment by nandorojo** [comment]
`useHref` seems logical to me. 

I _might_ prefer `asPath` over `href`, to avoid any confusion that `useHref().href` is the value passed to `<Link href />`

```tsx
<Link href={{ pathname: '/users/[id]', query: { id } }} />

useHref().href // i might expect this to be /users/[id] or something...
```

I might be overthinking this though.

**Comment by EvanBacon** [comment]
@nandorojo ya I'm not sold on `href` either, `asPath` is a bit confusing but might be better to unify. I'm not worried about devs passing `href.href` as it would technically always work:

```tsx
import { Link, useHref } from 'expo-router';

function App() {
  const href = useHref()
  return (
    <>
      <Link href={href}>Go nowhere</Link>
      <Link href={href.href}>Also go nowhere</Link>
    </>
  )
}
```

I went with `href` (after ~2.342s of thinking about it) because the type `Href` (`string | { pathname, query }`) used in `<Link href={...} />` is parsed from `{ pathname, query }` → `string`, implying that `href` the `string` would be the parsed results of `{ pathname, query }`. The reason we use `href` in `Link` instead of `to` (React Navigation, React Router) is because the pre-rendered/js-disabled results would be `<a href="..." />`. 

Some options:
- This same concept could kinda work if we changed to `<Link asPath={AsPath}>`.
- Maybe `useLocation` works. We wouldn't really be adding support for the fully qualified/parsed URL though (cuz we don't know which native URI scheme the user would want on native).
- Who cares about the uniformity, just rename it.

**Pull request merged**